File: MATTHEW/leetcode/AIcoding/5.Qwen-3/javascript_solutions/675-cut-off-trees-for-golf-event.js
function cutOffTree(land) {
    const m = land.length;
    const n = land[0].length;
    const queue = [];

    for (let i = 0; i < m; i++) {
        for (let j = 0; j < n; j++) {
            if (land[i][j] > 0) {
                queue.push([i, j]);
            }
        }
    }

    queue.sort((a, b) => land[a[0]][a[1]] - land[b[0]][b[1]]);

    let totalSteps = 0;
    let currentX = 0;
    let currentY = 0;

    while (queue.length > 0) {
        const [targetX, targetY] = queue.shift();
        const steps = bfs(land, currentX, currentY, targetX, targetY);
        if (steps === -1) return -1;
        totalSteps += steps;
        currentX = targetX;
        currentY = targetY;
    }

    return totalSteps;

    function bfs(grid, startX, startY, endX, endY) {
        const directions = [[-1, 0], [1, 0], [0, -1], [0, 1]];
        const visited = Array.from({ length: grid.length }, () => Array(grid[0].length).fill(false));
        const queue = [[startX, startY, 0]];
        visited[startX][startY] = true;

        while (queue.length > 0) {
            const [x, y, steps] = queue.shift();
            if (x === endX && y === endY) return steps;
            for (const [dx, dy] of directions) {
                const nx = x + dx;
                const ny = y + dy;
                if (nx >= 0 && nx < grid.length && ny >= 0 && ny < grid[0].length && !visited[nx][ny] && grid[nx][ny] > 0) {
                    visited[nx][ny] = true;
                    queue.push([nx, ny, steps + 1]);
                }
            }
        }

        return -1;
    }
}

File: leetcode/AIcoding/4.GPT-4o/java_solutions/675-cut-off-trees-for-golf-event.java
import java.util.*;

class Solution {
    public int cutOffTree(List<List<Integer>> forest) {
        int m = forest.size(), n = forest.get(0).size();
        List<int[]> trees = new ArrayList<>();
        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                if (forest.get(i).get(j) > 1) {
                    trees.add(new int[] {forest.get(i).get(j), i, j});
                }
            }
        }
        Collections.sort(trees, (a, b) -> Integer.compare(a[0], b[0]));

        int totalSteps = 0, startX = 0, startY = 0;
        for (int[] tree : trees) {
            int steps = bfs(forest, startX, startY, tree[1], tree[2]);
            if (steps == -1) return -1;
            totalSteps += steps;
            startX = tree[1];
            startY = tree[2];
        }
        return totalSteps;
    }

    private int bfs(List<List<Integer>> forest, int startX, int startY, int targetX, int targetY) {
        int m = forest.size(), n = forest.get(0).size();
        boolean[][] visited = new boolean[m][n];
        Queue<int[]> queue = new LinkedList<>();
        queue.offer(new int[] {startX, startY, 0});
        visited[startX][startY] = true;

        int[][] directions = {{1, 0}, {-1, 0}, {0, 1}, {0, -1}};

        while (!queue.isEmpty()) {
            int[] curr = queue.poll();
            int x = curr[0], y = curr[1], steps = curr[2];
            if (x == targetX && y == targetY) return steps;

            for (int[] dir : directions) {
                int newX = x + dir[0], newY = y + dir[1];
                if (newX >= 0 && newX < m && newY >= 0 && newY < n && forest.get(newX).get(newY) > 0 && !visited[newX][newY]) {
                    visited[newX][newY] = true;
                    queue.offer(new int[] {newX, newY, steps + 1});
                }
            }
        }
        return -1;
    }
}